tools/gcs-bucket-mover/gcs_bucket_mover/bucket_mover_service.py

```diff
@@ -0,0 +1,657 @@
+    bucket.lifecycle_rules = source_bucket_details.lifecycle_rules
+    bucket.versioning_enabled = source_bucket_details.versioning_enabled
+
+    if source_bucket_details.default_kms_key_name:
+        bucket.default_kms_key_name = source_bucket_details.default_kms_key_name
+        # The target project GCS service account must be given Encrypter/Decrypter permission for
+        # the key
+        _add_target_project_to_kms_key(
+            spinner, config, source_bucket_details.default_kms_key_name)
+
+    if source_bucket_details.logging:
+        bucket.enable_logging(source_bucket_details.logging['logBucket'],
+                              source_bucket_details.logging['logObjectPrefix'])
+
+    _create_bucket_api_call(spinner, bucket)
+
+    if source_bucket_details.iam_policy:
+        _update_iam_policies(config, bucket, source_bucket_details)
+        spinner.write(
+            'IAM policies successfully copied over from the source bucket')
+
+    if source_bucket_details.acl_entities:
+        new_acl = _update_acl_entities(config,
+                                       source_bucket_details.acl_entities)
+        bucket.acl.save(acl=new_acl)
+        spinner.write('ACLs successfully copied over from the source bucket')
+
+    if source_bucket_details.default_obj_acl_entities:
+        new_default_obj_acl = _update_acl_entities(
+            config, source_bucket_details.default_obj_acl_entities)
+        bucket.default_object_acl.save(acl=new_default_obj_acl)
+        spinner.write(
+            'Default Object ACLs successfully copied over from the source bucket'
+        )
+
+    if source_bucket_details.notifications:
+        _update_notifications(spinner, config,
+                              source_bucket_details.notifications, bucket)
+        spinner.write(
+            '{} Created {} new notifications for the bucket {}'.format(
+                CHECKMARK, len(source_bucket_details.notifications),
+                bucket_name))
+
+    return bucket
+
+
+def _retry_if_false(result):
+    """Return True if we should retry because the function returned False"""
+    return result is False
+
+
+@retry(
+    retry_on_result=_retry_if_false,
+    wait_exponential_multiplier=4000,
+    wait_exponential_max=60000,
+    stop_max_attempt_number=5)
+def _create_bucket_api_call(spinner, bucket):
+    """Calls the GCS api method to create the bucket.
+
+    The method will attemp to retry up to 5 times if the 503 ServiceUnavailable
+    exception is raised.
+
+    Args:
+        spinner: The spinner displayed in the console
+        bucket: The bucket object to create
+
+    Returns:
+        True if the bucket was created, False if a ServiceUnavailable exception was raised
+
+    Raises:
+        google.cloud.exceptions.Conflict: The underlying Google Cloud api will raise this error
+            if the bucket already exists.
+    """
+
+    try:
+        bucket.create()
+    except exceptions.ServiceUnavailable:
```

Comment:
The retrying module will retry on exception automatically. You're discarding the exception and adding in a retry function check which is really just throwing code at it for no reason, and losing some features.

This entire function should be:

bucket.create()
return True

I'd even suggest removing the return True line, the calling function can catch the exception (after the retries have happend) rather than check for a True value.

Reply:
In this case I want to handle and continue on a 503 (but only on a 503), and I want to let the user know what is happening if it occurs. I don't believe there's another way of doing this with retrying.

Reply:
My point is that discarding the exception and just returning a True/False value to the caller is not a great pattern. The caller cannot know if it's a timeout or something else went wrong. I admit that in this particular case that information won't be used but it's still a good practice.

Something like this snippet will still print notifications, still retry only on 503, but will also return the bucket.create() result to the caller, and it won't lose exception information if it's thrown.

```
def _retry_if_service_unavailable(exception):
  return isinstance(exception, exceptions.ServiceUnavailable)

@retry(
  retry_on_exception=_retry_if_service_unavailable,
  wait_exponential_multiplier=4000,
  wait_exponential_max=60000,
  stop_max_attempt_number=5)
def _create_bucket_api_call(spinner, bucket):
  try:
    return bucket.create()
  except exceptions.ServiceUnavailable:
    spinner.write('503 Service Unavailable error returned.'
                  ' Retrying up to 5 times with exponential backoff.')
    raise
```

Reply:
Ok, I see what you mean, I agree this looks like a better pattern.